content/2025-08-20-heapless-091.md

```diff
@@ -0,0 +1,114 @@
+
+So how can we obtain a `VecView` ? It's pretty simple: `Vec` can be *coerced* into a `VecView`. Coercion (in this case [`Unsized` coercion](https://doc.rust-lang.org/reference/type-coercions.html#r-coerce.unsize)), is a way the compiler can transform one type into another implicitly. In this case, the compiler is capable of converting pointers to a `Vec` (`&Vec<T, N>`, `&mut Vec<T, N>`, `Box<Vec<T, N>>` etc...) to pointers to a `VecView` (`&VecView<T>`, `&mut VecView<T>`, `Box<VecView<T>>` etc...), so you can use a reference to a `Vec` when a reference to a `VecView` is expected:
+
+```rust
+use heapless::{VecView, Vec};
+struct App{
+    …
+}
+
+impl App {
+     pub fn handle_request(input: &mut VecView<u8>, output: &mut Vec<u8>) -> Result<(), Error> {
+	     …
+     }
+}
+
+let mut request: Vec<u8, 256> = Vec::new();
+let mut reply: Vec<u8, 256> = Vec::new();
+
+app.handle_request(&mut request, &mut reply).unwrap();
+```
+
+If you prefer things to be explicit, the `View` variants of types (`Vec` is not the only data structure having `View` variants) can be obtained through `vec.as_view()` or through `vec.as_mut_view()`.
+
+The pointer to the `VecView` is the size of 2 `usize`: one for the address of the underlying `Vec`, and one for the capacity of the underlying `Vec`. This is exactly like slices. `VecView<T>` is to `Vec<T, N>` what a slice `[T]` is to an array `[T; N]`.
+Unless you need to store data on the stack, most often you will pass around `&mut [T]` rather than `&mut [T; N]`, because it's simpler. The same applies to `VecView`. Wherever you use `&mut Vec<T, N>`, you can instead use `&mut VecView<T>`.
+
+The `View` types are not available just for `Vec`. There are `View` versions of a lot of heapless types.
+
+## Benefits of the view types
+
+The benefits are multiple:
+
+### Better compatibility with `dyn Traits`
+
+If a trait has a function that takes a generic, it is not `dyn` compatible. By removing the const generic, the `View` types can make `dyn Trait` pass around data structures without having to hard-code a single size of buffer in the trait definition.
+
+### Better binary size and compile times
+
+When you use const-generics, the compiler needs to compile a new version of the function for each value of the const-generic.
+Removing the const generic means cutting down on duplicated functions that are all almost the same, which improves both compile time and the size of the resulting binary.
+
+### Better ergonomics
+
+The View types can remove a ton of excess noise from the generics.
+
+# The `LenType` optimization
+
+Most often, buffers in embedded applications will not contain a huge number of items.
+Until `0.9.1` the capacity of the `heapless` data structures were almost always stored as a `usize`, which can often encode much more values than necessary.
```

Comment:
capacity -> capacities

Reply:
I'm not sure. Capacities plural seems weird. I read it a the capacity of **each** of heapless data structures, even without the "each". But I'm not a native speaker either.